emu_sv/custom_callback_implementations.py

```diff
@@ -1,93 +1,117 @@
-import math
 import torch
 
-from emu_base.base_classes.config import BackendConfig
-from emu_base.base_classes.default_callbacks import (
-    QubitDensity,
-    EnergyVariance,
-    SecondMomentOfEnergy,
+from pulser.backend import (
     CorrelationMatrix,
+    EmulationConfig,
+    EnergySecondMoment,
+    EnergyVariance,
+    Occupation,
+    Energy,
 )
-from emu_base.base_classes.operator import Operator
 
-from emu_sv import StateVector
+from emu_sv.state_vector import StateVector
+from emu_sv.dense_operator import DenseOperator
 from emu_sv.hamiltonian import RydbergHamiltonian
 
 
-def qubit_density_sv_impl(
-    self: QubitDensity, config: BackendConfig, t: int, state: StateVector, H: Operator
+def qubit_occupation_sv_impl(
+    self: Occupation,
+    *,
+    config: EmulationConfig,
+    state: StateVector,
+    hamiltonian: DenseOperator,
 ) -> torch.Tensor:
     """
-    Custom implementation of the qubit density ❬ψ|nᵢ|ψ❭ for the state vector solver.
+    Custom implementation of the occupation ❬ψ|nᵢ|ψ❭ for the state vector solver.
     """
-    nqubits = int(math.log2(len(state.vector)))
-    state_tensor = state.vector.reshape((2,) * nqubits)
-
-    qubit_density = torch.zeros(nqubits, dtype=torch.float64, device=state_tensor.device)
+    nqubits = state.n_qudits
+    occupation = torch.zeros(nqubits, dtype=torch.float64, device=state.vector.device)
     for i in range(nqubits):
-        qubit_density[i] = state_tensor.select(i, 1).norm() ** 2
-    return qubit_density
+        state_tensor = state.vector.view(2**i, 2, -1)
+        # nᵢ is a projector and therefore nᵢ == nᵢnᵢ
+        # ❬ψ|nᵢ|ψ❭ == ❬ψ|nᵢnᵢ|ψ❭ == ❬ψ|nᵢ * nᵢ|ψ❭ == ❬ϕ|ϕ❭ == |ϕ|**2
+        occupation[i] = torch.linalg.vector_norm(state_tensor[:, 1]) ** 2
+    return occupation
 
 
 def correlation_matrix_sv_impl(
     self: CorrelationMatrix,
-    config: BackendConfig,
-    t: int,
+    *,
+    config: EmulationConfig,
     state: StateVector,
-    H: Operator,
+    hamiltonian: DenseOperator,
 ) -> torch.Tensor:
     """
-    Custom implementation of the density-density correlation ❬ψ|nᵢnⱼ|ψ❭ for the state vector solver.
-
+    Custom implementation of the density-density correlation ❬ψ|nᵢnⱼ|ψ❭
+      for the state vector solver.
     TODO: extend to arbitrary two-point correlation ❬ψ|AᵢBⱼ|ψ❭
     """
-    nqubits = int(math.log2(len(state.vector)))
-    state_tensor = state.vector.reshape((2,) * nqubits)
-
+    nqubits = state.n_qudits
     correlation = torch.zeros(
-        nqubits, nqubits, dtype=torch.float64, device=state_tensor.device
+        nqubits, nqubits, dtype=torch.float64, device=state.vector.device
     )
 
     for i in range(nqubits):
-        select_i = state_tensor.select(i, 1)
+        select_i = state.vector.view(2**i, 2, -1)
+        select_i = select_i[:, 1]
```

Comment:
you do know that line just calls `select_i.select(1,1)` under the hood right? And idem in all the other places where you replaced it. The reshape operation can be elided here.